### PR TITLE
amp-ad type=dotandads implementation

### PR DESCRIPTION
**Pull request by dotandads**
Provides support for amp-ad type=dotandads

### CODE REVIEW

**Comment by dotandads** [comment]
Company name: DotAndMedia S.r.l.

**Comment by dotandads** [comment]
I authored these commits

**Comment by giovanni-lovece-dm** [comment]
I authored these commits

**Comment by giovanni-lovece-dm** [comment]
I'm ok with these commits being contributed with Google.

**Comment by giovanni-lovece-dm** [comment]
How can we pass the cla check? We signed CLAs both as organization and as single person

**Comment by cramforce** [comment]
Thanks for the PR!

@dotandads Maybe you could try to comment again. Not sure how to do it. Ideally the author of the commit submits the PR. We will also later ask you to squash your commits to a single commit.
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

File: src/3p.js [hunk 1]
```diff
@@ -171,3 +171,18 @@ export function validateData(data, allowedFields) {
...
+
+/**
+ * Function to check if data has all mandatory
+ * fields specified
+ * @param {!Object} data
+ * @param {!Array<string>} fields
+ */
+export function dataHasAllFields(data, fields) {
+  for (let c = 0; c < fields.length; c++) {
+    if (!data.hasOwnProperty(fields[c]) && !data[fields[c]]) {
```
Comment (cramforce) [inline src/3p.js]: I think truthiness might be a bit too much for the general case. How about you do extended validation in the script you load?

File: ads/dotandads.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function dotandads(global, data) {
+  const mandatoryData = ["sp", "cid", "mpo", "mpt"];
+  if (dataHasAllFields(data, mandatoryData)) {
+    global.data = data;
+    writeScript(global, 'https://amp.ad.dotandad.com/dotandadsAmp.js');
+  } else {
+    log.error("DotAndAds", "missing mandatory fields");
```
Comment (cramforce) [inline ads/dotandads.js]: Please use `console.error` instead.

Use `console./*OK*/error` to tell the linter that you did not accidentally leave it in the code base

File: ads/dotandads.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    global.context.noContentAvailable();
```
Comment (cramforce) [inline ads/dotandads.js]: This is technically fine, but unless you expect errors to happen in prod I would not hide the slot.

**Comment by cramforce** [comment]
This looks great! Left some comments on the source.

File: src/3p.js (hunk 1 above)
Comment (dotandads) [inline src/3p.js]: Sure, we'll move the data validation in our script

File: ads/dotandads.js (hunk 3 above)
Comment (dotandads) [inline ads/dotandads.js]: As stated in a comment before, we'll move this check inside our loaded script

**Comment by dotandads** [comment]
Cool, we'll fix the code as suggested and we'll squash the commits together

**Comment by dotandads** [comment]
I've authored this commits

**Comment by cramforce** [comment]
@giovanni-lovece-dm Please say "I signed it"

**Comment by giovanni-lovece-dm** [comment]
I signed it

**Comment by dotandads** [comment]
I think I'll close this and reopen pull request with single commit

**Comment by ampsauce** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 4de2f8571aa72cecca400c8117e2e6238deb15b7
Build details: https://travis-ci.org/ampsauce/amphtml/builds/104935757

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by ampsauce** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 599f87cfa4d61a8aaf3aff9b94f1183a769bea09
Build details: https://travis-ci.org/ampsauce/amphtml/builds/104936509

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Pull request closed without merging**